src/encrypted_container.rs

```diff
@@ -0,0 +1,248 @@
+///
+/// Note: to decrease the CBOR overhead it might be useful to rename the serialized object fields for
+/// the serialization purposes. Use the `#[serde(rename = "A")]` attribute.
+///
+/// Usage example:
+/// ```compile_fail
+/// let ek = get_encryption_key(trussed)?;
+/// let data = EncryptedDataContainer::from_obj(trussed, obj, None, ek)?;
+/// let data_serialized: Message = data.try_into()?;
+/// ```
+/// Future work and extensions:
+/// - Generalize over serialization method
+/// - Generalize buffer size
+/// - Investigate postcard structure extensibility, as a means for smaller overhead for serialization
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct EncryptedDataContainer {
+    #[serde(rename = "D")]
+    data: trussed::types::Message,
+    #[serde(rename = "T")]
+    tag: trussed::types::ShortData,
+    #[serde(rename = "N")]
+    nonce: trussed::types::ShortData,
+}
+
+impl TryFrom<&[u8]> for EncryptedDataContainer {
+    type Error = Error;
+
+    /// Create an instance from this serialized Encrypted Data Container
+    fn try_from(value: &[u8]) -> Result<Self> {
+        let credential =
+            cbor_deserialize(value).map_err(|_| Error::DeserializationToContainerError)?;
+        Ok(credential)
+    }
+}
+
+impl TryFrom<EncryptedDataContainer> for Message {
+    type Error = Error;
+
+    /// Try to serialize EncryptedDataContainer to Bytes
+    fn try_from(value: EncryptedDataContainer) -> Result<Self> {
+        let mut buf = [0u8; SERIALIZED_OBJECT_BUFFER_SIZE];
+        let r = cbor_serialize(&value, &mut buf).map_err(|_| Error::ObjectSerializationError)?;
+        let bytes = Message::from_slice(r).map_err(|_| Error::SerializationBufferTooSmall)?;
+        Ok(bytes)
+    }
+}
+
+impl EncryptedDataContainer {
+    /// Decrypt given Bytes and return original object instance
+    pub fn decrypt_from_bytes<T, O>(
+        trussed: &mut T,
+        ser_encrypted: Message,
+        encryption_key: KeyId,
+    ) -> Result<O>
+    where
+        T: trussed::Client + trussed::client::Chacha8Poly1305,
+        for<'a> O: Deserialize<'a>,
+    {
+        let deserialized_container: EncryptedDataContainer =
+            cbor_deserialize(&ser_encrypted).map_err(|_| Error::DeserializationToContainerError)?;
+
+        deserialized_container.decrypt(trussed, None, encryption_key)
+    }
+
+    /// Create Encrypted Data Container from the given object
+    pub fn from_obj<T, O>(
+        trussed: &mut T,
+        obj: &O,
+        associated_data: Option<&[u8]>,
+        encryption_key: KeyId,
+    ) -> Result<EncryptedDataContainer>
+    where
+        T: trussed::Client + trussed::client::Chacha8Poly1305,
+        O: Serialize,
+    {
+        let mut buf = [0u8; SERIALIZED_OBJECT_BUFFER_SIZE];
+        let message = cbor_serialize(obj, &mut buf).map_err(|_| Error::ObjectSerializationError)?;
```

Comment:
`Message::try_from` could be used here too.

Reply:
Here I am passing a raw slice, so does not help entirely, but removing the buffer type  will be an improvement anyway, as it will implementation depending on the Trussed's Message, and removing type size indirection. Done as suggested.